language_service/src/display.rs

```diff
@@ -0,0 +1,573 @@
+        };
+
+        write!(
+            f,
+            "{} {} {} {} {}{}",
+            kind,
+            self.decl.name.name,
+            HirTy {
+                ty: &self.decl.input.ty,
+                compilation: self.compilation
+            },
+            arrow,
+            HirTy {
+                ty: &self.decl.output,
+                compilation: self.compilation
+            },
+            FunctorSetValue {
+                functors: self.decl.functors,
+            },
+        )
+    }
+}
+
+pub(crate) struct AstCallableDecl<'a> {
+    compilation: &'a Compilation,
+    decl: &'a ast::CallableDecl,
+}
+
+impl<'a> Display for AstCallableDecl<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        let (kind, arrow) = match self.decl.kind {
+            ast::CallableKind::Function => ("function", "->"),
+            ast::CallableKind::Operation => ("operation", "=>"),
+        };
+
+        let functors = ast_callable_functors(self.decl);
+        let functors = FunctorSetValue { functors };
+
+        write!(
+            f,
+            "{} {} {} {} {}{}",
+            kind,
+            self.decl.name.name,
+            TyId {
+                ty_id: self.decl.input.id,
+                compilation: self.compilation
+            },
+            arrow,
+            Ty {
+                ty: &self.decl.output
+            },
+            functors,
+        )
+    }
+}
+
+pub(crate) struct IdentTyDef<'a> {
+    ident: &'a ast::Ident,
+    def: &'a ast::TyDef,
+}
+
+impl<'a> Display for IdentTyDef<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{} = {}", self.ident.name, TyDef { def: self.def })
+    }
+}
+
+pub(crate) struct HirUdt<'a> {
+    compilation: &'a Compilation,
+    ident: &'a hir::Ident,
+    udt: &'a hir::ty::Udt,
+}
+
+impl<'a> Display for HirUdt<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        let udt_def = UdtDef::new(self.compilation, self.udt);
+        write!(f, "{} = {}", self.ident.name, udt_def)
+    }
+}
+
+struct UdtDef<'a> {
+    compilation: &'a Compilation,
+    name: Option<Rc<str>>,
+    kind: UdtDefKind<'a>,
+}
+
+enum UdtDefKind<'a> {
+    SingleTy(&'a hir::ty::Ty),
+    TupleTy(Vec<UdtDef<'a>>),
+}
+
+impl<'a> UdtDef<'a> {
+    pub fn new(compilation: &'a Compilation, def: &'a hir::ty::Udt) -> Self {
+        let mut udt_def = UdtDef::convert_hir_ty_to_udt(compilation, &def.base);
+        for field in &def.fields {
+            UdtDef::update_udt_def(&mut udt_def, &field.name, &field.path.indices);
+        }
+        udt_def
+    }
+
+    fn convert_hir_ty_to_udt(compilation: &'a Compilation, ty: &'a hir::ty::Ty) -> UdtDef<'a> {
+        match ty {
+            hir::ty::Ty::Tuple(tys) => UdtDef {
+                compilation,
+                name: None,
+                kind: UdtDefKind::TupleTy(
+                    tys.iter()
+                        .map(|ty| UdtDef::convert_hir_ty_to_udt(compilation, ty))
```

Comment:
Is it expected that these nested `UdtDef`s will _not_ have `update_udt_def` called on them and therefore will all have `name: None`?

Reply:
That's right. The base type first gets converted into a UdtDef struct with no names by this function. Then update_udt_def is called with the field names to put the names in the correct places in the UdtDef struct.